java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -267,37 +290,49 @@ public void completeShard(@NonNull final ShardedJobRunId jobId, @NonNull final I
   private <T extends IncrementalTask> IncrementalTaskState<T> getAndValidateTaskState(Transaction tx, IncrementalTaskId taskId,
                                                                                       int sequenceNumber, ShardedJobStateImpl<T> jobState) {
     IncrementalTaskState<T> taskState = lookupTaskState(tx, taskId);
-    if (taskState == null) {
-      log.warning(taskId + ": Task gone");
+    IncrementalTaskStateValidationError validationError = validateTaskState(taskState, sequenceNumber, jobState);
+    if (validationError != null) {
       return null;
+    } else {
+      return taskState;
+    }
+  }
+
+  enum IncrementalTaskStateValidationError {
+    TASK_GONE,
+    TASK_NO_LONGER_ACTIVE,
+    JOB_NO_LONGER_ACTIVE,
+    TASK_STATE_FROM_PAST,
+    TASK_ALREADY_COMPLETED,
+    LOCK_HELD_BY_OTHER_EXECUTION,
+    ;
```

Comment:
maybe add a TASK_OK, rather than deal with NULL, works as default too. and rename to  `IncrementalTaskStateStatus`

Reply:
how about `IncrementalTaskStateState`? 👅